Make sure that EventHandler only tracks target nodes in

ziransun opened the pull request:
#### fd0fbd758e5e66ddf886ffe774a0b05995ca2a87
<pre>
Make sure that EventHandler only tracks target nodes in
its document.

<a href="https://bugs.webkit.org/show_bug.cgi?id=234977">https://bugs.webkit.org/show_bug.cgi?id=234977</a>

Reviewed by Darin Adler and Ryosuke Niwa.

When an element is removed from the document, we should clear up
m_lastElementUnderMouse.

* LayoutTests/TestExpectations:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateMouseEventTargetNode):
* LayoutTests/fast/events/mouseover-mouseout2-expected.txt:
* LayoutTests/fast/events/mouseover-mouseout2.html:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_from_slot_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_lostpointercapture_for_disconnected_node-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_reappending_last_over_target-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_mouse-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_pen-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_mouse-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_pen-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_from_slot_mouse-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_mouse-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_pen-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=LEFT&amp;button=0&amp;buttons=1-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&amp;button=1&amp;buttons=4-expected.txt:

Canonical link: <a href="https://commits.webkit.org/285250@main">https://commits.webkit.org/285250@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/32527c75d62c6b695ca24b4d53d8512a03c73166

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71858 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51278 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24643 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76015 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23067 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73973 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59079 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22887 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56731 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15227 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74925 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46520 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61911 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37170 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43183 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19381 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21412 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65094 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19745 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77698 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16098 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18933 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64788 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16144 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61934 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64459 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15909 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12628 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6275 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47076 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1860 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48145 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49432 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47889 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2794,17 +2794,17 @@ void EventHandler::updateMouseEventTargetNode(const AtomString& eventType, Node*
+        m_lastElementUnderMouse = nullptr;
+        m_lastScrollbarUnderMouse = nullptr;
+    }
     ASSERT_IMPLIES(m_lastElementUnderMouse, &m_lastElementUnderMouse->document() == frame->document());
```
This assertion seems a bit silly since it’s now the same as the code above.

darinadler commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2794,17 +2794,17 @@ void EventHandler::updateMouseEventTargetNode(const AtomString& eventType, Node*
 
     ASSERT_IMPLIES(m_elementUnderMouse, &m_elementUnderMouse->document() == frame->document());
+
+    if (m_lastElementUnderMouse && &m_lastElementUnderMouse->document() != frame->document()) {
```
This seems like it’s happening too late. It should have happened right when the `m_lastElementUnderMouse` was moved to a new document, not waited until this function was called.

ziransun replied:
Updated the code. PTAL. Thank you!

darinadler commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -2346,6 +2346,11 @@ HandleUserInputEventResult EventHandler::handleMouseReleaseEvent(const PlatformM
+    if (m_lastElementUnderMouse && &m_lastElementUnderMouse->document() != frame->document()) {
+        m_lastElementUnderMouse = nullptr;
+        m_lastScrollbarUnderMouse = nullptr;
+    }
```
We could have cleared out `m_lastElementUnderMouse` at the time the element moved, rather than waiting until `EventHandler::handleMouseReleaseEvent` is called. There is a `Document` function called when an element is removed from a document, and this might be efficient enough to call there.

ziransun replied:
Thanks very much for the suggestion! Code updated. PTAL.

rniwa commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -486,6 +486,11 @@ void EventHandler::nodeWillBeRemoved(Node& nodeToBeRemoved)
+
+    if (nodeToBeRemoved.containsIncludingShadowDOM(m_lastElementUnderMouse.get())) {
+        m_lastElementUnderMouse = nullptr;
+        m_lastScrollbarUnderMouse = nullptr;
```
Are we sure m_lastScrollbarUnderMouse is always inside m_lastElementUnderMouse?

ziransun replied:
No, I'm not sure. How do we check it? My intention  is to remove  m_lastElementUnderMouse but I noticed that at https://searchfox.org/wubkat/source/Source/WebCore/page/EventHandler.cpp#2805, m_lastScrollbarUnderMouse was removed at the same time. Maybe I should leave m_lastScrollbarUnderMouse  out since I'm not sure?

rniwa replied:
Leaving that out for now seems like a good move.

ziransun replied:
Thanks! Code updated.

darinadler commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -486,6 +486,9 @@ void EventHandler::nodeWillBeRemoved(Node& nodeToBeRemoved)
     if (nodeToBeRemoved.containsIncludingShadowDOM(RefPtr { m_clickNode }.get()))
         m_clickNode = nullptr;
+
+    if (nodeToBeRemoved.containsIncludingShadowDOM(m_lastElementUnderMouse.get()))
```
Why doesn’t this have the `RefPtr { }.get()` the way the similar code above does?

rniwa replied:
We probably need to do that. Safer C++ check is gonna fail otherwise.

ziransun committed:
Make sure that EventHandler only tracks target nodes in
its document.

https://bugs.webkit.org/show_bug.cgi?id=234977

Reviewed by Darin Adler and Ryosuke Niwa.

When an element is removed from the document, we should clear up
m_lastElementUnderMouse.

* LayoutTests/TestExpectations:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateMouseEventTargetNode):
* LayoutTests/fast/events/mouseover-mouseout2-expected.txt:
* LayoutTests/fast/events/mouseover-mouseout2.html:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_from_slot_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_lostpointercapture_for_disconnected_node-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_reappending_last_over_target-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_mouse-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_interleaved.tentative_pen-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_mouse-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_appended_pen-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_from_slot_mouse-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_mouse-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_after_target_removed_interleaved.tentative_pen-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/mouse_boundary_events_after_removing_last_over_element-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=LEFT&button=0&buttons=1-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/uievents/mouse/synthetic-mouse-enter-leave-over-out-button-state-after-target-removed.tentative_buttonType=MIDDLE&button=1&buttons=4-expected.txt:

Canonical link: https://commits.webkit.org/285250@main
